### ISSUE TITLE
[Version 2.0] Make it possible to generate QR-Codes on a website

### ISSUE THREAD

**Issue by Ein-Tim** [corona-warn-app/cwa-wishlist#399]
## Current Implementation
QR-Codes can only be generated inside of the app ([Source](https://github.com/corona-warn-app/cwa-documentation/issues/554#issuecomment-800922574)).

## Suggested Enhancement
Add a new section to www.coronawarn.app which enables users to generate QR-Codes online on the website.

## Expected Benefits
Better user experience, it's easier to go through the steps on a large monitor than on a little phone screen. It's also easier to print QR-Codes out if the printer is connected via cable. 

## Related Issue
https://github.com/corona-warn-app/cwa-documentation/issues/554

---
Internal Tracking ID: [EXPOSUREAPP-5880](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-5880)

**Comment by KST1810** [comment]
I like the idea! But I believe that this would require for the data of the CWA to be stored on a centralized server, right?

**Comment by Ein-Tim** [comment]
@KST1810 

No, if I understand the [documentation](https://github.com/corona-warn-app/cwa-documentation/blob/master/event_registration.md), which was published today right, this shouldn't be necessary.

The QR code can be generated in app with [this](https://github.com/corona-warn-app/cwa-documentation/blob/master/event_registration.md#qr-code-structure-draft) information.

**Comment by dsarkar** [comment]
@Ein-Tim Thanks.

---
Internal Tracking ID: [EXPOSUREAPP-5880](https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-5880)

**Comment by fynngodau** [comment]
I imagine somebody who needs to generate a lot of QR codes would find it very uncomfortable to do that on their mobile devices. For example, a local transport company might want to equip their bus with QR codes. It would be much more comfortable to do that on a computer with a keyboard, a proper "batch creation" feature and the ability to print all of the codes at once.

**Comment by KST1810** [comment]
I think batch creation sounds great! 

But couldn't we also create a "parent node" that sits in the location at all times and works just like a normal node and tracks guests etc?

**Comment by fynngodau** [comment]
@KST1810 That sounds architectionically challanging at least. However, your suggestion has inspired me to think about alternative check-in methods: #406

**Comment by KST1810** [comment]
@fynngodau may I ask why?

Couldn't the CWA simply use a normal phone and assign it the parameter "host-node"? Then everything would be tracked, wouldn't it?

**Comment by jucktnich** [comment]
The problem is, it would be hard to get a detail screen, with information about the event.

**Comment by Ein-Tim** [comment]
@jucktnich 

Are you talking about the printing out or about the NFC solution? 
If it's about the printing out, why would it be hard to get this screen?

**Comment by jucktnich** [comment]
It's about the parent node solution.

**Comment by fynngodau** [comment]
> But couldn't we also create a "parent node" that sits in the location at all times and works just like a normal node and tracks guests etc?

The way you phrased it, it sounds like this node would recursively those who have seen it once it itself has seen one user who caused a risk warning. However it cannot be verified whether this is actually true.

The other way round could work somehow I guess, but it would be technically challenging. And it would somewhat defeat the point of creating a new system that works in parallel to the bluetooth tracing and specifically in situations where guests might breathe the same air but not be at close distance.

Anyhow this is quite off-topic for this issue, so we should probably stop discussing this here (and open a new issue if required).

**Comment by pkreissel** [comment]
Another possibility would be to use this website on a tablet or other screen on the entrance to the venue. This could enable the owners to frequently and automatically change the QR-Code displayed there, which would prevent the "Profiling of specific venues" attack mentioned in the [documentation](https://github.com/corona-warn-app/cwa-documentation/blob/master/event_registration.md).

**Comment by KST1810** [comment]
@pkreissel on that note: 

As plenty of restaurants, pubs etc have TV's or similar displays, would there be a chance to create an "CWA QR-Code display-app" that automatically iterates through a number of codes to prevent the aforementioned attacks?

**Comment by Ein-Tim** [comment]
@KST1810 

Feel free to open another issue reg. this since such an app is not really the topic of this issue.

**Comment by pinguinpfleger** [comment]
> on a tablet or other screen on the entrance to the venue.

I join the wish.
At my store door I have a DoorEpaper (e-ink ePaper on a Pi Zero W) in operation.
A possibility to regenerate the QR code via a cron job would be a reliable solution.

**Comment by MaZderMind** [comment]
I implemented a Python3-Version of the Event-QR-Code-Spec which can be used to generate QR-Codes in an offline, bulk or other automated scenario: https://github.com/MaZderMind/cwa-qr

Note that I was unable to End-to-End-Test the generated Codes, because this would require a positive Covid-Test. Use at your own risk (as far as that's possible, these days…).

**Comment by thomasaugsten** [comment]
@MaZderMind To generate a valid QR-code you have to add the cryptographicSeed and the publicKey
The publicKey is a fix key for all QR-Codes


You can find an example here:
https://github.com/corona-warn-app/cwa-documentation/blob/master/event_registration.md#qr-code-structure

**Comment by MaZderMind** [comment]
@thomasaugsten I pulled the Public-Key from the OpenSource iOS App and am generating a Random Seed as required to avoid Location Tracking, see https://github.com/MaZderMind/cwa-qr/blob/09863c7/cwa.py#L66-L67 – is this incorrect? If so, let's open an Issue in https://github.com/MaZderMind/cwa-qr/issues and keep this discussion clean.

**Comment by thomasaugsten** [comment]
Looks good

**Comment by dsarkar** [comment]
We would be interested in more details, which features, e.g.:

- Create batch of QR codes at the same time, e.g. for a chain of stores
- Create QR codes for one or more places using a schedule/calendar (Day of the week and times)
- .... more ideas ?

**Comment by fkusei** [comment]
In my case, we do have some displays running a digital signage solution, which are displaying the latest news from our intranet. I want to integrate a CWA compatible QR code into that, so employees will be able to check in to "their" display. This allows us to roll out a large number of QR codes without needing to print them out and hang them individually.

**Comment by dsarkar** [comment]
Dear all, 

a community member is already developing a QR code generator for the website, see https://github.com/corona-warn-app/cwa-website/pull/1130. Of course we will evaluate how to proceed with this proposal and contributions. 

We have the following question to you: 
As the CWA project is a decentralized approach, data and privacy are highly protected. Being able to create QR codes on the CWA website, what do think would be the **perception** with regard to data and privacy protection ?

**Comment by Ein-Tim** [comment]
Thank you for asking @dsarkar!

The data should only be processed on the local machine, not on any server. I this is true, I think that there should be an additional paragraph which informs the user where the data is processed. 
Then everything will be fine I think.

**Comment by mtb77** [comment]
I added some ideas to the PR https://github.com/corona-warn-app/cwa-website/pull/1130 such as rolling displays and batch creation to use the QR code generation on different digital signage or tablet devices

**Comment by fynngodau** [comment]
> Being able to create QR codes on the CWA website, what do think would be the perception with regard to data and privacy protection ?

As @Ein-Tim already mentioned, there could be a hint to the fact that the data is not stored by the server.

I personally would make it look like a little info box with a padlock icon, containing the following text: "**Secure QR code generation** – QR codes are processed exclusively on your device, and only you can access QR codes you create."

**Comment by dsarkar** [comment]
FYI https://www.coronawarn.app/en/blog/2021-04-23-qr-code-generator/

**Comment by Ein-Tim** [comment]
Thank you! Closing as implemented! Really nice!

**Comment by dsarkar** [comment]
Dear community, @bastianjoel, @Ein-Tim, and many others to mention ...

Thanks for proposing, commenting, proving a PR, and contributing in general. Great work, and great collaboration!

---
Corona-Warn-App Open Source Team

**Comment by Ein-Tim** [comment]
@dsarkar 

Thank you also for the good cooperation! And a huge thanks to the responsible people for this quick approval! 🎉
